refactor: introduced UTC+0000 for timing task check

zhouzilong2020 opened the pull request:
After this PR, timing task check will store, compare time in UTC+0000 at the backend. 
And at the frontend, I added some prompts to inform users' local timezone.
For readability, users are still allowed to pick a date in their local time zone, but we will convert it to UTC+0000 before sending an API request to the backend.

zhouzilong2020 commented:
PTAL

zhouzilong2020 commented on server/task_check_executor_timing.go:
```diff
@@ -32,24 +32,27 @@ func (exec *TaskCheckTimingExecutor) Run(ctx context.Context, server *Server, ta
-	if payload.EarliestAllowedTs == 0 {
+	// we use UTC+0000 time for comparison
+	utc, _ := time.LoadLocation(time.UTC.String())
+	// EarliestAllowedTs is store as UTC+0000
```
```suggestion
	// EarliestAllowedTs is store in UTC+0000
```

tianzhou commented on api/task.go:
```diff
@@ -131,11 +131,12 @@ type Task struct {
+	Status  TaskStatus `jsonapi:"attr,status"`
+	Type    TaskType   `jsonapi:"attr,type"`
+	Payload string     `jsonapi:"attr,payload"`
+	// This field is stored in UTC+0000
```
I guess it's OK to not mention this since our convention is always storing the allowed ts in UTC

zhouzilong2020 replied:
```suggestion
```

tianzhou commented on frontend/src/utils/util.ts:
```diff
@@ -59,6 +59,26 @@ export function timezoneString(zoneName: string, offset: number): string {
   ).padStart(2, "0")}`;
 }
 
+export const UTC0ToLocal = (utc0Ts: number | null): number | null => {
```
UTCToLocal, same in other places.

no need to mention 0 if we are just using UTC+0

tianzhou commented on server/task_check_executor_timing.go:
```diff
@@ -43,13 +43,16 @@ func (exec *TaskCheckTimingExecutor) Run(ctx context.Context, server *Server, ta
 	}
 
-	if time.Now().Before(time.Unix(payload.EarliestAllowedTs, 0)) {
+	// we use UTC+0000 time for comparison
```
can we simply compare between time.Unix() and payload.EarliestAllowedTs ?

zhouzilong2020 replied:
Not likely. time.Unix() would return local time (see https://pkg.go.dev/time#Now)

zhouzilong2020 commented on frontend/src/utils/util.ts:
```diff
@@ -59,6 +59,26 @@ export function timezoneString(zoneName: string, offset: number): string {
+export const UTC0ToLocal = (utc0Ts: number | null): number | null => {
+  const diff = dayjs().utcOffset();
+  const localTs = utc0Ts
+    ? dayjs(utc0Ts * 1000)
```
I guess this awkwardness is caused by the inconsistency between the backend and the frontend. 
Most frontend frameworks (moment.js, day.js, naive-UI) would take timestamps in nanoseconds, there are dozens of *1000 spreading all over the project.

zhouzilong2020 commented:
PTAL

zhouzilong2020 commented on frontend/src/utils/util.ts:
```diff
@@ -59,6 +59,36 @@ export function timezoneString(zoneName: string, offset: number): string {
+ * @returns local timestamp in milisecond
+ */
+export const UTCToLocal = (utcTs: number | null): number | null => {
+  const diff = dayjs().utcOffset();
```
I meant to mean second. It was a miswrite.
Already corrected the comments, PTAL.
(In all, the frontend would prefer timestamp in milliseconds, and backend would prefer timestamp in second.)

zhouzilong2020 commented on frontend/src/components/Issue/IssueSidebar.vue:
```diff
@@ -104,8 +110,12 @@
             (newTimestampNs) => {
-              state.earliestAllowedTs = newTimestampNs / 1000;
-              $emit('update-earliest-allowed-time', newTimestampNs / 1000);
+              // n-date-picker would pass timestamp in milisecond.
```
As it is described in [2], Unix time is a single signed number that increments every second. 
However, in my environments (both safari and chrome) console.log('Barny1: ', newTimestampNs) would yield '1642126472000', which is clearly a millisecond unit.

zhouzilong2020 replied:
Oh, I may misread this. I got your point.

zhouzilong2020 commented:
PTAL

zhouzilong2020 commented:
> Any change? Let us know when it's ready. Thx!

Removed the conversion at the frontend and changed the timezone to UTC at the backend when prompting error message

zhouzilong2020 commented:
PTAL

tianzhou reviewed (approved):
LGTM
